src/main/kotlin/com/cjbooms/fabrikt/generators/controller/KtorControllerInterfaceGenerator.kt

```diff
@@ -0,0 +1,362 @@
+            }
+        }
+
+        bodyParams.forEach { param ->
+            codeBlock.appendLine("val ${param.name} = call.receive<${param.type.simpleName()}>()")
+        }
+
+        val methodParameters =
+            listOf(headerParams, pathParams, queryParams, bodyParams).flatten()
+                .joinToString(", ") { it.name }
+                .let {
+                    if (it.isNotEmpty()) "call, $it"
+                    else "call"
+                }
+
+        codeBlock.appendLine("val result = controller.$methodName($methodParameters)")
+
+        if (happyPathResponse.simpleName() == Unit::class.simpleName) {
+            // When return type is Unit we only respond with the status code.
+            // Note however that in some cases the controller may choose to respond directly on the call
+            // in which case that takes precedence in Ktor's routing. For example: call.respondRedirect().
+            codeBlock.appendLine("call.respond(result.status)")
+        } else {
+            codeBlock.appendLine("call.respond(result.status, result.message)")
+        }
+
+        codeBlock.appendLine("}")
+
+        if (addAuth) {
+            codeBlock.appendLine("}")
+        }
+
+        return codeBlock.toString()
+    }
+
+    override fun generateLibrary(): Collection<ControllerLibraryType> = listOf(
+        buildRoutingUtils(),
+        buildControllerResult()
+    )
+
+    /**
+     * Builds a data class to be used as a return type for controller methods
+     */
+    private fun buildControllerResult(): ControllerLibraryType {
+        val genericT = TypeVariableName("T")
+
+        val spec = TypeSpec.classBuilder(CONTROLLER_RESULT_CLASS_NAME)
+            .addModifiers(KModifier.DATA)
+            .addTypeVariable(genericT)
+            .primaryConstructor(
+                FunSpec.constructorBuilder()
+                    .addParameter("status", ClassName("io.ktor.http", "HttpStatusCode"))
+                    .addParameter("message", genericT) // naming aligned with Ktor's call.response(status, message)
+                    .build()
+            )
+            .addProperty(
+                PropertySpec.builder("status", ClassName("io.ktor.http", "HttpStatusCode"))
+                    .initializer("status")
+                    .build()
+            )
+            .addProperty(
+                PropertySpec.builder("message", genericT)
+                    .initializer("message")
+                    .build()
+            )
+            .build()
+
+        return ControllerLibraryType(spec, packages.base)
+    }
+
+    /**
+     * Builds a utility object to be used for parsing query parameters
+     */
+    private fun buildRoutingUtils(): ControllerLibraryType {
+        val genericR = TypeVariableName("R", Any::class)
+
+        val returnType = genericR.copy(nullable = true, reified = true)
+
+        val spec = TypeSpec.objectBuilder("RoutingUtils")
```

Comment:
I don't quite see why this should be a `public object` instead of some `private fun`ctions. Foremost, this feels like it could cause name clashes within a package if you have multiple controller files. Also, it might clutter the import space.

Reply:
The two functions `Parameters.getTyped` and `Headers.getOrFail` are used by routes where parameter and header parsing is performed, and therefore contained in a utility object. 

The object is only generated once through the `generateLibrary` function.

They are compliments to the built in Ktor server function [Parameters.getOrFail](https://api.ktor.io/ktor-server/ktor-server-core/io.ktor.server.util/[jvm-and-nix]get-or-fail.html).

Reply:
> The two functions Parameters.getTyped and Headers.getOrFail are used by routes where parameter and header parsing is performed, and therefore contained in a utility object.
> The object is only generated once through the generateLibrary function.

Yeah I get the use case. I just meant that I wouldn't have used a wrapping object which afaik doesn't serve any purpose here other than syntactic structure. But I understand that might just be a stylistic preference, so I don't want to argue against it.

> They could also be private functions on the controller interfaces, but they would then be replicated across all controllers. But perhaps that's OK?

I do believe private functions would be better (i.e. private free functions, or functions within a private object). Otherwise they become part of the public interface given by Fabrikt, which we probably don't want (Hyrum's law :) )

Reply:
Excellent point about not polluting the public API of the generated code. I have made the two functions part of the controller interface's companion object which allows them to be used by the route function without polluting the API.